Mathlib/GroupTheory/Perm/Centralizer.lean

```diff
@@ -0,0 +1,646 @@
+      (comm := fun a ha b hb h ↦ g.pairwise_commute_of_mem_zpowers h (v a) (v b) (v a).2 (v b).2) hx
+    exact support_zpowers_of_mem_cycleFactorsFinset_le (v a) ha
+
+theorem θHom_injective (g : Perm α) : Function.Injective (θHom g) := by
+  rw [θHom, MonoidHom.noncommCoprod_injective]
+  refine ⟨ofSubtype_injective, ?_, ?_⟩
+  · apply injective_noncommPiCoprod_of_independent
+    intro a
+    simp only [zpowers_eq_closure, ← closure_iUnion]
+    apply disjoint_closure_of_disjoint_support
+    rintro - ⟨-⟩ - ⟨-, ⟨b, rfl⟩, -, ⟨h, rfl⟩, ⟨-⟩⟩
+    rw [← disjoint_iff_disjoint_support]
+    exact cycleFactorsFinset_pairwise_disjoint g a.2 b.2 (mt Subtype.ext_iff.mpr h.symm)
+  · rw [noncommPiCoprod_range, ← ofSubtype.range.closure_eq]
+    simp only [zpowers_eq_closure, ← closure_iUnion]
+    apply disjoint_closure_of_disjoint_support
+    rintro - ⟨a, rfl⟩ - ⟨-, ⟨b, rfl⟩, ⟨-⟩⟩
+    exact (ofSubtype_support_disjoint a).mono_right (mem_cycleFactorsFinset_support_le b.2)
+
+theorem θHom_range_eq :
+    (θHom g).range = (toPermHom g).ker.map (Subgroup.subtype _) := by
+  apply le_antisymm
+  · rw [θHom, MonoidHom.noncommCoprod_range, sup_le_iff, noncommPiCoprod_range, iSup_le_iff]
+    simp only [zpowers_le]
+    constructor
+    · rintro - ⟨a, rfl⟩
+      refine ⟨⟨ofSubtype a, ?_⟩, ?_, rfl⟩
+      · rw [mem_centralizer_singleton_iff]
+        exact Disjoint.commute (disjoint_iff_disjoint_support.mpr (ofSubtype_support_disjoint a))
+      · exact Perm.ext fun x ↦ Subtype.ext (disjoint_iff_disjoint_support.mpr
+          ((ofSubtype_support_disjoint a).mono_right
+            (mem_cycleFactorsFinset_support_le x.2))).commute.mul_inv_cancel
+    · intro i
+      refine ⟨⟨i, mem_centralizer_singleton_iff.mpr (self_mem_cycle_factors_commute i.2)⟩, ?_, rfl⟩
+      exact Perm.ext fun x ↦ Subtype.ext (cycleFactorsFinset_mem_commute' g i.2 x.2).mul_inv_cancel
+  · rintro - ⟨p, hp, rfl⟩
+    simp only [coeSubtype]
+    set u : Perm (Function.fixedPoints g) :=
+      subtypePerm p (fun x ↦ mem_fixedPoints_iff_apply_mem_of_mem_centralizer p.2)
+    simp only [SetLike.mem_coe, mem_ker_toPermHom_iff, IsCycle.forall_commute_iff] at hp
+    set v : (c : g.cycleFactorsFinset) → (Subgroup.zpowers (c : Perm α)) :=
+      fun c => ⟨ofSubtype
+          (p.1.subtypePerm (Classical.choose (hp c.val c.prop))),
+            Classical.choose_spec (hp c.val c.prop)⟩
+    use (u, v)
+    ext x
+    rw [θHom_apply]
+    split_ifs with hx
+    · rw [cycleOf_mem_cycleFactorsFinset_iff, mem_support] at hx
+      rw [ofSubtype_apply_of_mem, subtypePerm_apply]
+      rwa [mem_support, cycleOf_apply_self, ne_eq]
+    · rw [cycleOf_mem_cycleFactorsFinset_iff, not_mem_support] at hx
+      rwa [ofSubtype_apply_of_mem, subtypePerm_apply]
+
+theorem θHom_range_card (g : Equiv.Perm α) :
+    Fintype.card (θHom g).range = (Fintype.card α - g.cycleType.sum)! * g.cycleType.prod := by
+  erw [Set.card_range_of_injective (θHom_injective g)]
```

Comment:
I suggest adding these lemmas to `Mathlib.Algebra.Group.Subgroup.Finite` and using it instead of the erw:
```
lemma _root_.Fintype.card_coeSort_mrange {M N : Type*} [Monoid M] [Monoid N] [Fintype M]
    [DecidableEq N] {f : M →* N} (hf : Function.Injective f) :
    Fintype.card (mrange f) = Fintype.card M :=
  Set.card_range_of_injective hf

lemma _root_.Fintype.card_coeSort_range [Fintype G] [DecidableEq N] {f : G →* N}
    (hf : Function.Injective f) :
    Fintype.card (range f) = Fintype.card G :=
  Set.card_range_of_injective hf
```

Reply:
Just noting that we already have the `Nat.card` version `Nat.card_range_of_injective`.